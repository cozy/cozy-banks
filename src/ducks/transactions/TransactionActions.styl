@require 'settings/breakpoints'

.TransactionAction
    cursor pointer
    text-decoration none
    // need to reset the padding otherwise since the link
    // in a menu has a padding to account for the background-icon
    // TODO remove this padding from the ActionMenu
    padding 0 !important
    white-space inherit !important

.TransactionAction-disabled
    cursor: default

.TransactionActionMenu
    width 100%

    +medium-screen()
        text-align center

.TransactionActionMenuItem
    text-align left
