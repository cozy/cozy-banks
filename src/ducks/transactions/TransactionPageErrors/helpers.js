import orderBy from 'lodash/orderBy'
import { trigger as triggerLibs } from 'cozy-client/dist/models'

const { triggers: triggerUtils, triggerStates } = triggerLibs

/**
 * Returns errors that should be displayed in the error container
 *
 * @returns
 */
export const getTriggersOrderByError = ({
  triggers,
  accounts,
  isBankTrigger
}) => {
  const accountsConnectionId = accounts
    .map(account => account.relationships?.connection?.data?._id)
    .filter(Boolean)

  const bankTriggers = triggers
    .filter(triggerUtils.isKonnectorWorker)
    .filter(isBankTrigger)
    .filter(trigger => accountsConnectionId.includes(trigger.message.account))

  const bankTriggersOrdered = orderBy(
    bankTriggers,
    triggerStates.isErrored,
    'desc'
  )

  return bankTriggersOrdered
}
