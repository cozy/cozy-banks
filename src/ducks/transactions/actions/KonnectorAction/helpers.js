export function getBrandsWithoutTrigger(brands) {
  return brands.filter(brand => !brand.hasTrigger)
}
