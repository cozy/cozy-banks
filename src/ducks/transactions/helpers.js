export const getLabel = transaction =>
  transaction.label.toLowerCase().replace(/(?:^|\s)\S/g, a => a.toUpperCase())
