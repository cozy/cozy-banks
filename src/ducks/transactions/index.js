import TransactionsPage, {
  TransactionsPageWithBackButton
} from 'ducks/transactions/TransactionsPage'
export { getLabel } from 'ducks/transactions/helpers'

// components
export { TransactionsPage, TransactionsPageWithBackButton }
