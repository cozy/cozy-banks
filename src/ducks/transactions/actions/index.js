import { find, filter } from 'lodash'
import KonnectorAction from './KonnectorAction'
import HealthLinkAction from './HealthLinkAction'
import AppLinkAction from './AppLinkAction'
import UrlLinkAction from './UrlLinkAction'
import BillAction from './BillAction'
import AlertAction from './AlertAction'
import CommentAction from './CommentAction'
import AttachAction from './AttachAction'
import HealthExpenseAction from './HealthExpenseAction'
import HealthExpenseStatusAction from './HealthExpenseStatusAction'

const actions = [
  HealthExpenseStatusAction,
  HealthExpenseAction,
  HealthLinkAction,
  KonnectorAction,
  UrlLinkAction,
  BillAction,
  AppLinkAction,
  AttachAction,
  CommentAction,
  AlertAction
]

export const findMatchingActions = async (transaction, actionProps) => {
  const matchingActions = []

  for (const action of actions) {
    const matching = await action.match(transaction, actionProps)
    if (matching) {
      matchingActions.push(action)
    }
  }

  const defaultAction = find(
    matchingActions,
    action => action.defaultAction !== false && action.disabled !== true
  )
  const othersAction = filter(
    matchingActions,
    action => action !== defaultAction
  )

  return {
    default: defaultAction,
    others: othersAction
  }
}

export const getActionFromName = name => {
  return find(actions, action => action.name === name)
}
