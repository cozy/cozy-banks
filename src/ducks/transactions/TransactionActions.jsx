/**
 * Is used in both TransactionActionMenu and TransactionMenu
 * to show possible actions related to a transaction.
 *
 * The TransactionAction (the action the user is most susceptible
 * to need) can also be shown in desktop mode, directly in the
 * table.
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { MenuItem, Icon } from 'cozy-ui/react'
import palette from 'cozy-ui/stylus/settings/palette.json'
import { findMatchingActions } from './actions'

// TODO delete or rename this variable (see https://gitlab.cozycloud.cc/labs/cozy-bank/merge_requests/237)
const PRIMARY_ACTION_COLOR = palette.dodgerBlue

const MenuIcon = ({ action, transaction, actionProps, color }) => {
  if (action.getIcon) {
    return action.getIcon({ action, transaction, actionProps })
  }
  return <Icon icon={action.icon} color={color} />
}

const MenuAction = ({
  action,
  transaction,
  actionProps,
  className,
  isDefault,
  displayComponent = true,
  compact = false,
  menuPosition,
  onlyItems
}) => {
  const { Component } = action
  const color = action.disabled
    ? palette.charcoalGrey
    : action.color || actionProps.color || PRIMARY_ACTION_COLOR
  return (
    <MenuItem
      className={className}
      disabled={action.disabled}
      onClick={actionProps.onClick}
      icon={
        !isDefault && (
          <MenuIcon
            action={action}
            transaction={transaction}
            actionProps={actionProps}
            color={color}
          />
        )
      }
    >
      {displayComponent && (
        <Component
          action={action}
          transaction={transaction}
          actionProps={actionProps}
          color={color}
          compact={compact}
          menuPosition={menuPosition}
          onlyItems={onlyItems}
        />
      )}
    </MenuItem>
  )
}

MenuAction.propTypes = {
  actionProps: PropTypes.object.isRequired
}

export const SyncTransactionActions = ({
  transaction,
  actions,
  actionProps,
  displayDefaultAction,
  onlyDefault,
  onlyIcon,
  menuPosition = 'left',
  onlyItems,
  compact
}) => (
  <span>
    {(displayDefaultAction || onlyDefault) &&
      actions.default && (
        <MenuAction
          action={actions.default}
          className={onlyDefault || onlyIcon ? 'u-p-0' : ''}
          displayComponent={!onlyIcon}
          isDefault
          transaction={transaction}
          actionProps={actionProps}
          menuPosition={menuPosition}
          onlyItems={onlyItems}
          compact={compact}
        />
      )}
    {!onlyDefault &&
      actions.others.map((action, index) => (
        <MenuAction
          key={index}
          action={action}
          transaction={transaction}
          actionProps={actionProps}
        />
      ))}
  </span>
)

class TransactionActions extends Component {
  state = {
    actions: false,
    actionProps: false
  }

  async findMatchingActions() {
    const { transaction } = this.props
    if (transaction) {
      const { urls, brands, bill } = this.props
      const actionProps = { urls, brands, bill }
      const actions = await findMatchingActions(transaction, actionProps)
      this.setState({ actions, actionProps })
    }
  }

  componentDidMount() {
    this.findMatchingActions()
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.transaction !== this.props.transaction) {
      this.setState({ actions: false })
    }
  }

  componentDidUpdate(nextProps) {
    if (
      nextProps.transaction !== this.props.transaction ||
      nextProps.urls !== this.props.urls ||
      nextProps.brands !== this.props.brands ||
      nextProps.bill !== this.props.bill
    ) {
      this.findMatchingActions()
    }
  }

  render() {
    const { actions, actionProps } = this.state

    if (!actions) return null

    return (
      <SyncTransactionActions
        actions={actions}
        actionProps={actionProps}
        {...this.props}
      />
    )
  }
}

TransactionActions.propTypes = {
  transaction: PropTypes.object.isRequired,
  urls: PropTypes.object.isRequired,
  brands: PropTypes.array.isRequired
}

export default TransactionActions
