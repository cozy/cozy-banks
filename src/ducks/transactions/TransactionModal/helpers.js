import Alerter from 'cozy-ui/transpiled/react/deprecated/Alerter'

import { getDate, getApplicationDate } from 'ducks/transactions/helpers'

export const showAlertAfterApplicationDateUpdate = (transaction, t, f) => {
  const date = getApplicationDate(transaction) || getDate(transaction)
  Alerter.success(
    t('Transactions.infos.applicationDateChangedAlert', {
      applicationDate: f(date, 'MMMM')
    })
  )
}

export const stopPropagation = ev => ev.stopPropagation()
