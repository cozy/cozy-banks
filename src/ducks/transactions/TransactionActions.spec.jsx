import dummyjson from 'dummy-json'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'
import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import merge from 'lodash/merge'
import keyBy from 'lodash/keyBy'

import { I18n } from 'cozy-ui/react'
import { SyncTransactionActions } from './TransactionActions'
import { hydrateReimbursementWithBill } from 'documents/transaction'
import { findMatchingActions } from 'ducks/transactions/actions'
import langEn from 'locales/en.json'

import dataTpl from '../../../test/fixtures/operations.json'
import helpers from '../../../test/fixtures/helpers'
import brands from 'ducks/brandDictionary/brands'

configure({ adapter: new Adapter() })

const data = JSON.parse(
  dummyjson.parse(
    JSON.stringify(dataTpl),
    merge(helpers, {
      helpers: {
        reference: x => x
      }
    })
  )
)

const store = createStore(() => ({
  mobile: {
    url: 'cozy-url://'
  }
}))

jest.mock('utils/documentCache', () => ({
  get: (doctype, id) => mockBillsById[id]
}))

// Find a way to better do that, here we have to copy the definition
// of isProperIcon
jest.mock('cozy-ui/react/Icon', () => {
  const isProperIcon = icon => {
    const isSvgSymbol = icon && !!icon.id
    const isIconIdentifier = typeof icon === 'string'
    return isSvgSymbol || isIconIdentifier
  }
  const mockIcon = props => {
    const icon = props.icon
    return isProperIcon(icon) ? <span data-icon-id={icon.id || icon} /> : icon
  }
  mockIcon.isProperIcon = isProperIcon
  return mockIcon
})

const bills = data['io.cozy.bills']
// prefixed by mock to be used inside documentCache mock
const mockBillsById = keyBy(bills, x => x._id)
const transactions = data['io.cozy.bank.operations']
const getBill = billId => mockBillsById[billId]
const hydratedTransactions = transactions.map(x => ({
  ...x,
  bill: mockBillsById[x.billId],
  reimbursements:
    x.reimbursements &&
    x.reimbursements.map(r => hydrateReimbursementWithBill(r, getBill))
}))
const transactionsById = keyBy(hydratedTransactions, '_id')

/* eslint-disable */
const tests = [
  // transaction id, class variant, text, icon, action name, [action props], [test name]
  ['paiementdocteur', null, '2 reimbursements', 'file', 'HealthExpenseStatus'],
  ['paiementdocteur2', '.c-actionbtn--error', 'No reimbursement yet', 'hourglass', 'HealthExpenseStatus'],
  ['depsantelou1', '.c-actionbtn--error', 'No reimbursement yet', 'hourglass', 'HealthExpenseStatus'],
  ['depsantegene4', '.c-actionbtn--error', 'No reimbursement yet', 'hourglass', 'HealthExpenseStatus'],
  ['depsanteisa2', '.c-actionbtn--error', 'No reimbursement yet', 'hourglass', 'HealthExpenseStatus'],
  ['depsantecla3', '.c-actionbtn--error', 'No reimbursement yet', 'hourglass', 'HealthExpenseStatus'],
  ['facturebouygues', null, '1 invoice', 'file', 'bill'],
  ['salaireisa1', null, 'Accéder à votre paie', 'openwith', 'url'],
  ['fnac', null, 'Accéder au site Fnac', 'openwith', 'url'],
  ['edf', null, 'EDF', null, 'app'],
  ['remboursementcomplementaire', null, 'My invoices', 'plus', 'konnector', {
    brands: brands.filter(x => x.name == 'Malakoff Mederic')
  }, 'remboursementcomplementaire konnector not installed'],
  ['remboursementcomplementaire', null, '1 invoice', null, 'bill', {
    brands: []
  }, 'remboursementcomplementaire konnector installed'],
  ['normalshopping', null, 'toto', null]
]
/* eslint-enable */

const AppLike = ({ children }) => (
  <I18n lang={'en'} dictRequire={() => langEn}>
    <Provider store={store}>{children}</Provider>
  </I18n>
)

const actionProps = {
  urls: {
    COLLECT: 'collect',
    edf: 'edf-url://',
    maif: 'maifurl'
  },

  // Brands represents the brands for which no konnector
  // has been installed
  brands: brands.filter(x => x.name == 'Malakoff Mederic')
}

describe('transaction action defaults', () => {
  for (let test of tests) {
    const [
      id,
      variant,
      text,
      icon,
      actionName,
      specificActionProps = {},
      extraName = ''
    ] = test
    describe(`${id} ${extraName}`, () => {
      let root, actions
      beforeAll(async () => {
        const transaction = transactionsById[test[0]]
        const mergedActionProps = { ...actionProps, ...specificActionProps }
        actions = await findMatchingActions(transaction, mergedActionProps)
        root = mount(
          <AppLike>
            <SyncTransactionActions
              onlyDefault
              transaction={transaction}
              actions={actions}
              actionProps={mergedActionProps}
            />
          </AppLike>
        )
      })

      if (actionName) {
        it('should render the correct action', () => {
          expect(actions.default).not.toBe(undefined)
          expect(actions.default.name).toBe(actionName)
        })
      } else {
        it('should not render an action', () => {
          expect(actions.default).toBe(undefined)
        })
      }

      if (actionName) {
        it('should render the correct text', () => {
          root.update() // https://github.com/airbnb/enzyme/issues/1233#issuecomment-340017108
          const btn = root.find('.c-actionbtn')
          const btnText = btn.text()
          expect(btnText).toEqual(expect.stringContaining(text))
        })
      }

      if (icon) {
        it('should render the correct icon', () => {
          expect(root.find(`[data-icon-id="${icon}"]`).length).toBe(1)
        })
      }

      if (variant) {
        it('should render the correct button', () => {
          expect(root.find(variant).length).toBe(1)
        })
      }
    })
  }
})

describe('transaction action menu', () => {
  for (let test of tests) {
    const [id, , , , , specificActionProps = {}, extraName = ''] = test
    it(`should render correctly ${id} ${extraName}`, async () => {
      const transaction = transactionsById[id]
      const mergedActionProps = { ...actionProps, ...specificActionProps }
      const actions = await findMatchingActions(transaction, mergedActionProps)
      const root = mount(
        <AppLike>
          <SyncTransactionActions
            transaction={transaction}
            actions={actions}
            actionProps={mergedActionProps}
          />
        </AppLike>
      )
      const texts = root.find('.TransactionAction').map(x => x.text())
      expect(texts).toMatchSnapshot()
    })
  }
})
