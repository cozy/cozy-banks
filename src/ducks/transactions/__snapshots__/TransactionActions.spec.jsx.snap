// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction action menu should render correctly depsantecla3  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly depsantegene4  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly depsanteisa2  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly depsantelou1  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly edf  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly facturebouygues  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly fnac  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly normalshopping  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly paiementdocteur  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly paiementdocteur2  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly remboursementcomplementaire remboursementcomplementaire konnector installed 1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly remboursementcomplementaire remboursementcomplementaire konnector not installed 1`] = `
Array [
  "Bill",
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;

exports[`transaction action menu should render correctly salaireisa1  1`] = `
Array [
  "Attach a document",
  "Add a comment",
  "Create an alert",
]
`;
